src/app/tests/TestReadInteraction.cpp

```diff
@@ -1319,10 +1319,11 @@ void TestReadInteraction::TestSetDirtyBetweenChunks(nlTestSuite * apSuite, void
 
         ctx.DrainAndServiceIO();
 
-        // We should receive another 2 * (6 + 1) = 14 attribute reports since the underlying path iterator should be reset.
+        // We should receive another (6 + 1) = 7 attribute reports since the underlying path iterator should be reset to the
+        // beginning of the cluster it is current iterating.
         ChipLogError(DataManagement, "OLD: %d\n", currentAttributeResponsesWhenSetDirty);
         ChipLogError(DataManagement, "NEW: %d\n", delegate.mNumAttributeResponse);
-        NL_TEST_ASSERT(apSuite, delegate.mNumAttributeResponse == currentAttributeResponsesWhenSetDirty + 14);
+        NL_TEST_ASSERT(apSuite, delegate.mNumAttributeResponse == currentAttributeResponsesWhenSetDirty + 7);
```

Comment:
Can we restructure these tests to:

```
Mark Dirty (Item1, Item2, Item ...)
Generate Report Series()
Mark Dirty (Item4)
Validate Receiver (Item1, Item2, Item 3)
Generate Report Series()
Validate Receiver (Item4)
```

or something like this. This kind of macro-style, programming-instruction approach makes it easier to validate these various corner cases.

Right now, this logic is going based purely on counts, which is not at all accurate, and very brittle.

Reply:
I'd like to see tests for the following (A, B, C = Clusters, 1, 2, 3 = Attribute IDs):

Basic:
1. Mark A1, A2, A3 dirty and deliver A1, A2, A3 in a single report.
2. "" + Mark A1, A2 again dirty, and validate we get A1, A2.

Advanced:
3. Mark A1 A2 A3 dirty, and deliver A1, A2 in a single chunked report. Validate we got A1, A2. Mark A2 dirty again, and generate another chunked report with A2 in it, and validate we get the newer A2, A3.
4. Mark A2 A2 A3 dirty, and deliver A1, A2 in one chunked report. Mark A1 dirty again. Validate we get A3 in the next chunked report, and then a new report chunks with A1 in it.
5. Any other tests you can think of.

To do the above, you'll need to be able to artificially limit the number of attributes that can be packed in a report in the reporting engine during test, as well as being able to tag the report chunking series (i.e exchanges) by assigning a number during OnReportBegin on the client.

Reply:
Infact this is not what this test meant for, I have added another test in `controller/tests/TestReadChunking.cpp` and it should covered these.